### ISSUE TITLE
jj git clone fails with ecdsa-sk hardware SSH key

### ISSUE THREAD

**Issue by emesterhazy** [jj-vcs/jj#2958]
## Description

`jj git clone git@github.com:martinvonz/jj.git` fails if the user is using an SSH key with a
hardware security key such as the types described in Github's [SSH key documentation](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent#generating-a-new-ssh-key-for-a-hardware-security-key).

## Steps to Reproduce the Problem

1. Create a new SSH key with `ssh-keygen -t ecdsa-sk -C "your_email@example.com"` and a hardware security key.
1. Run `jj git clone git@github.com:martinvonz/jj.git`.
1. See an error and the repo is not cloned.

## Expected Behavior

The repo should be cloned.

## Actual Behavior

```
jj git clone git@github.com:martinvonz/jj.git
Error: authentication required but no callback set; class=Ssh (23); code=Auth (-16)
Hint: Jujutsu uses libssh2, which doesn't respect ~/.ssh/config. Does `ssh -F /dev/null` to the host work?
```

## Specifications

- Platform: MacOs
- Version: `jj 0.13.0`

**Comment by martinvonz** [comment]
Same as #2931?

**Comment by emesterhazy** [comment]
Sorry, the title was incorrect. It said ed25519-sk but this issue is actually for ecdsa-sk. In either case I think this is potentially different since it's the hardware security key version of the key. If fixing #2931 somehow fixes this issue as well that would be great though.

**Comment by necauqua** [comment]
While this might be a libssh2 issue or something else, I want to mention that gpg-backed ssh keys work great :) incl. with a gpg keycard impl, such as yubikey (me) and presumably others

**Comment by gdonval** [comment]
> Same as #2931?

That looks like a general configuration problem: OP reported RSA doesn't work either.

**Comment by Nemo157** [comment]
I've done a little investigation as I can't push using an `ed25519-sk` key either. 

<details><summary>probably irrelevant packaging debugging to check that there is -sk support at all</summary>

From what I can see libssh2 added support for -sk in [v1.11.0](https://libssh2.org/changes.html), the latest version of the `libssh2-sys` crate vendors [libssh2 v1.10.0](https://docs.rs/crate/libssh2-sys/latest/source/libssh2/include/libssh2.h) and by default [uses the vendored version instead of system version](https://docs.rs/crate/libssh2-sys/latest/source/build.rs), but I'm using a nixpkgs build which [disables the vendored libssh2](https://github.com/NixOS/nixpkgs/blob/nixos-23.11/pkgs/applications/version-management/jujutsu/default.nix#L37), _but_ this silently fails because of a missing `zlib` system package and fallsback to using the vendored version (silent until I added `-vv` to get log output from the build script), but even after patching that and confirming I have a version of `jj` built against `libssh2 v1.11.0` it still fails.

</details>

Running with more logs I see that it is trying the ssh-agent first, and I see my security key request activation, but then presumably that fails to authenticate for some reason, then it sees the ssh-key and tries to use it but fails because no passphrase is passed

```
2024-03-02T00:15:57.702907Z DEBUG run:run_internal:run_command: git2::cred: executing credential helper "sh" "-c" "git credential-cache get"
2024-03-02T00:15:57.708990Z TRACE run:run_internal:run_command: git2::cred: credential helper stderr ---

2024-03-02T00:15:57.709008Z TRACE run:run_internal:run_command: git2::cred: ignoring output line:
2024-03-02T00:15:57.709048Z  INFO run:run_internal:run_command: jj_lib::git: trying ssh_key_from_agent username="git"
2024-03-02T00:16:02.431408Z DEBUG run:run_internal:run_command: git2::cred: executing credential helper "sh" "-c" "git credential-cache get"
2024-03-02T00:16:02.436531Z TRACE run:run_internal:run_command: git2::cred: credential helper stderr ---

2024-03-02T00:16:02.436539Z TRACE run:run_internal:run_command: git2::cred: ignoring output line:
2024-03-02T00:16:02.436588Z  INFO run:run_internal:run_command:get_ssh_keys{_username="git"}: jj_cli::git_util: found ssh key path="/home/nemo157/.ssh/id_ed25519_sk"
2024-03-02T00:16:02.436601Z  INFO run:run_internal:run_command: jj_lib::git: trying ssh_key username="git" path="/home/nemo157/.ssh/id_ed25519_sk"
Error: Failed to authenticate SSH session: Unable to extract public key from private key file: Wrong passphrase or invalid/unrecognized private key file format; class=Ssh (23)
```

**Comment by Nemo157** [comment]
I built Jujutsu against a version of libssh2 with debug logging enabled <details><summary>(nix overlay)</summary>

```nix
pkgs-prev.jujutsu.overrideAttrs {
  buildInputs = with pkgs-final; [
    openssl
    zstd
    libgit2
    zlib
    (libssh2.overrideAttrs {
      configureFlags = ["--enable-debug"];
      patches = libssh2.patches ++ [
        (writeText "log-everything.patch" ''
diff --git a/src/misc.c b/src/misc.c
index 7de6511b54...105a821b2f 100644
--- a/src/misc.c
+++ b/src/misc.c
@@ -526,11 +526,6 @@
     const char *contexttext = contexts[0];
     unsigned int contextindex;

-    if(!(session->showmask & context)) {
-        /* no such output asked for */
-        return;
-    }
-
     /* Find the first matching context string for this message */
     for(contextindex = 0; contextindex < ARRAY_SIZE(contexts);
          contextindex++) {
        '')
      ];
    })
  ];
}
```

</details>

From those logs it appears to successfully send the sign-request to the ssh-agent, then forward the result to the server, which responds with an authentication failure. I also dumped the sign-request from the agent-side and compared to the sign-request sent by `git`, afaict they are using identical parameters, so it seems like that part is fine. I guess the next step would be to setup a test sshd-instance and try and see what's different about the auth request sent to the server that makes it get rejected.

**Comment by bnjmnt4n** [comment]
Are your keys passphrase-protected, and are you able to use/using the ssh-agent to store passphrases?

**Comment by Nemo157** [comment]
Yes, they're passphrase protected, which appears to be why the subsequent attempt to use the key directly fails; they're stored unlocked in ssh-agent.

**Comment by bnjmnt4n** [comment]
Yeah, I think the libgit2/libssh2 integration isn't the best; I don't have the most experience with the technology. If you'd like, you can also try https://github.com/martinvonz/jj/pull/3191 which uses OpenSSH directly for creating the connection, which should probably work better with the ssh-agent.

**Comment by Nemo157** [comment]
Yep, that branch appears to work fine. It can both use the key from the agent, and use it from disk and auto-load it into the agent.

**Comment by lilydjwg** [comment]
My jj git clone fails with an ed25519_sk key being present; the key is not the correct one to use but jj doesn't try with the rest even it has found it.

**Comment by bsdinis** [comment]
Can we close this now that #5228 is merged?

**Comment by emesterhazy** [comment]
Thanks for going back and closing issues obsoleted by the new git subprocess implementation.

As a gratuitous post hoc comment, I agree with the git subprocess approach being the most pragmatic and flexible. I suspect many companies and teams have non-vanilla authentication processes (i.e. sso, security keys like this issue mentions) that they've already configured to work with the git binary. Delegating to git instead of trying to implement feature parity frees up resources for other exciting jj feature work.
